jumanji/environments/combinatorial/cvrp/env.py

```diff
@@ -0,0 +1,285 @@
+# limitations under the License.
+
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+from chex import PRNGKey
+from jax import random
+
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.environments.combinatorial.cvrp.specs import ObservationSpec
+from jumanji.environments.combinatorial.cvrp.types import Observation, State
+from jumanji.environments.combinatorial.cvrp.utils import (
+    DEPOT_IDX,
+    compute_tour_length,
+    generate_problem,
+)
+from jumanji.types import Action, TimeStep, restart, termination, transition
+
+
+class CVRP(Environment[State]):
+    """
+    Capacitated Vehicle Routing Problem (CVRP) environment as described in [1].
+    - observation: Observation
+        - coordinates: jax array (float32) of shape (num_nodes + 1, 2)
+            the coordinates of each node and the depot
+        - demands: jax array (float32) of shape (num_nodes + 1,)
+            the associated cost of each node and the depot (0.0 for the depot)
+        - position: jax array (int32)
+            the index of the last visited node
+        - capacity: jax array (float32)
+            the current capacity of the vehicle
+        - action_mask: jax array (bool) of shape (num_nodes + 1,)
+            binary mask (False/True <--> invalid/valid action)
+
+    - reward: jax array (float32)
+        the negative sum of the distances between consecutive nodes at the end of the episode (the
+        reward is 0 if a previously selected non-dept node is selected again, or the depot is
+        selected twice in a row)
+
+    - state: State
+        - coordinates: jax array (float32) of shape (num_nodes + 1, 2)
+            the coordinates of each node and the depot
+        - demands: jax array (int32) of shape (num_nodes + 1,)
+            the associated cost of each node and the depot (0.0 for the depot)
+        - position: jax array (int32)
+            the index of the last visited node
+        - capacity: jax array (int32)
+            the current capacity of the vehicle
+        - visited_mask: jax array (bool) of shape (num_nodes + 1,)
+            binary mask (False/True <--> not visited/visited)
+        - order: jax array (int32) of shape (2 * num_nodes,)
+            the identifiers of the nodes that have been visited (-1 means that no node has been
+            visited yet at that time in the sequence)
+        - num_visits: int32
+            number of actions that have been taken (i.e., unique visits)
+
+    [1] Toth P., Vigo D. (2014). "Vehicle routing: problems, methods, and applications".
+    """
+
+    def __init__(
+        self, problem_size: int = 100, max_capacity: int = 30, max_demand: int = 10
+    ):
+        assert max_capacity >= max_demand, (
+            f"The demand associated with each node must be lower than the maximum capacity, "
+            f"hence the maximum capacity must be >= {max_demand}."
+        )
```

Comment:
this should be raised as an error instead of an assert. Asserts should only be used for debugging purposes as they can be turned off. Anything that is "user error" like creating an Env with capacity > demand should be an raised as an error :)